### DISCUSSION TITLE
Is there a way to decompile a single function/method call?

### DISCUSSION THREAD

**Discussion by compasslg** [icsharpcode/ILSpy#2226] [Help] [answered]
Hey friends! I looked online, and it was suggested to use ICSharpCode.Decompiler.Ast.AstBuilder to do it, but it seems like this class no longer exist. So is there a replacement for it? tbh there are few documentation and up to date example to it, which sucks.

It would be best to decompile a single method call (which should be possible considering how ILSpy displays c# code and IL code together), but I'll accept decompiling an entire method too.

Any ideas?

**Comment by dgrunwald** [comment, marked as answer]
`class AstBuilder` was replaced by `class CSharpDecompiler` in ILSpy 3.0.

You can decompile a single method using `CSharpDecompiler.Decompile()`, passing in the `MethodDefinitionHandle`.
The "C# with IL" view works by decompiling the whole method, generating sequence points (debug information) for the decompiled code; and then using those sequence points to map the generated statements back to the IL instructions.
https://github.com/icsharpcode/ILSpy/blob/master/ILSpy/Languages/CSharpILMixedLanguage.cs

**Reply by compasslg** [reply]
Thank you! But how am I supposed to get the MethodDefinitionHandle out of a MethodDefinition?

**Reply by dgrunwald** [reply]
Use [MetadataTokens.EntityHandle](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.metadata.ecma335.metadatatokens.entityhandle?view=netcore-2.2) to convert a metadata token into a handle.
`var handle = (MethodDefinitionHandle)MetadataTokens.EntityHandle(cecilMethod.MetadataToken.ToInt32())`

See [this wiki page](https://github.com/icsharpcode/ILSpy/wiki/srm) for more information on the switch from Cecil to System.Reflection.Metadata.
